sdks/python/apache_beam/ml/transforms/embeddings/open_ai.py

```diff
@@ -0,0 +1,223 @@
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# OpenAI Python SDK is required for this module.
+# Install using: pip install openai
+
+import logging
+import time
+from collections.abc import Iterable
+from collections.abc import Sequence
+from typing import Any, Optional, TypeVar, Union, Tuple, cast
+
+import apache_beam as beam
+from apache_beam.pvalue import PCollection, Row
+from apache_beam.io.components.adaptive_throttler import AdaptiveThrottler
+from apache_beam.metrics.metric import Metrics
+from apache_beam.ml.inference.base import ModelHandler
+from apache_beam.ml.inference.base import RunInference
+from apache_beam.ml.transforms.base import EmbeddingsManager
+from apache_beam.ml.transforms.base import _TextEmbeddingHandler
+from apache_beam.utils import retry
+
+__all__ = ["OpenAITextEmbeddings"]
+
+# Define a type variable for the output
+MLTransformOutputT = TypeVar('MLTransformOutputT')
+
+# Default batch size for OpenAI calls
+_BATCH_SIZE = 20  # OpenAI can handle larger batches than Vertex
+_MSEC_TO_SEC = 1000
+
+LOGGER = logging.getLogger("OpenAIEmbeddings")
+
+
+def _retry_on_appropriate_openai_error(exception):
+  """
+    Retry filter that returns True if a returned error is rate limit (429) or server error (5xx).
+
+    Args:
+        exception: the returned exception encountered during the request/response
+          loop.
+
+    Returns:
+        boolean indication whether or not the exception is a Server Error (5xx) or
+          a RateLimitError (429) error.
+    """
+  return isinstance(exception, (RateLimitError, APIError))
+
+
+class _OpenAITextEmbeddingHandler(ModelHandler):
+  """
+    Note: Intended for internal use and guarantees no backwards compatibility.
+    """
+
+  def __init__(
+      self,
+      model_name: str,
+      api_key: Optional[str] = None,
+      organization: Optional[str] = None,
+      dimensions: Optional[int] = None,
+      user: Optional[str] = None,
+  ):
+    """Initialize OpenAI embedding handler.
+
+        Args:
+            model_name: Name of the OpenAI embedding model (e.g., "text-embedding-3-small")
+            api_key: OpenAI API key
+            organization: OpenAI organization ID
+            dimensions: Specific embedding dimensions to use (if model supports it)
+            user: End-user identifier for tracking and rate limit calculations
```

Comment:
Check the formatting/linting guide for how to automate fixing formatting - https://cwiki.apache.org/confluence/display/BEAM/Python+Tips#PythonTips-LintandFormattingChecks